stable_baselines3/her/her_replay_buffer.py

```diff
@@ -0,0 +1,193 @@
+            her_new_goals = [trans["achieved_goal"] for trans in random_transitions]
+        else:
+            raise ValueError("Strategy for sampling goals not supported!")
+
+        # assign new goals as desired_goals
+        for idx, goal in enumerate(her_new_goals):
+            transitions[her_idxs][:, 0][idx]["desired_goal"] = goal
+
+        observations, actions, rewards, next_observations, dones = list(zip(*transitions))
+
+        # compute new rewards with new goal
+        achieved_goals = [new_obs["achieved_goal"] for new_obs in np.array(next_observations)[her_idxs]]
+        new_rewards = np.array(rewards)
+        new_rewards[her_idxs] = [
+            self.env.env_method("compute_reward", ag, her_new_goals, None)
+            for ag, new_goal in zip(achieved_goals, her_new_goals)
+        ]
+
+        # concatenate observation with (desired) goal
+        obs = [np.concatenate([o["observation"], o["desired_goal"]], axis=1) for o in observations]
+        new_obs = [np.concatenate([new_o["observation"], new_o["desired_goal"]], axis=1) for new_o in next_observations]
+
+        data = (
+            np.array(obs)[:, 0, :],
+            np.array(actions, dtype=self.action_space.dtype)[:, 0, :],
+            np.array(new_obs)[:, 0, :],
+            np.array(dones, dtype=np.int8),
+            new_rewards,
+        )
+
+        return ReplayBufferSamples(*tuple(map(self.to_torch, data)))
+
+    def add(self, obs: np.ndarray, next_obs: np.ndarray, action: np.ndarray, reward: np.ndarray, done: np.ndarray) -> None:
+        """
+        Add episode to replay buffer
+
+        :param obs:
+        :param next_obs:
+        :param action:
+        :param reward:
+        :param done:
+
+        :param episode: (list) Episode to store.
+        """
+        episode = list(zip(obs, action, reward, next_obs, done))
+
+        episode_length = len(episode)
+
+        # check if replay buffer has enough space for all transitions of episode
+        if self.n_transitions_stored + episode_length <= self.size():
+            self.buffer.append(episode)
+            # update replay size
+            self.current_size += 1
+            self.n_transitions_stored += episode_length
+        elif self.full:
+            # if replay buffer is full take random stored episode and replace it
+            idx = np.random.randint(0, self.current_size)
+
+            if len(self.buffer[idx]) == episode_length:
+                self.buffer[idx] = episode
+            elif len(self.buffer[idx]) > episode_length:
+                self.buffer[idx] = episode
+                self.n_transitions_stored -= self.buffer[idx] - episode_length
+
+        if self.n_transitions_stored == self.size():
+            self.full = True
+        else:
+            self.full = False
+
+    def get_current_episode_size(self):
+        return self.current_size
+
+    def get_current_size(self):
+        return self.n_transitions_stored
+
+    def get_transitions_stored(self):
```

Comment:
Maybe make  `self.n_transitions_stored` private and create a getter using `@property`